Source/Mothership/SuperDB.cpp

```diff
@@ -0,0 +1,92 @@
+#include "SuperDB.h"
+
+/* Cleanup. */
+SuperDB::~SuperDB()
+{
+    for (SuperIt superIt = supervisors.begin(); superIt != supervisors.end();
+         superIt++) delete superIt->second;
+    supervisors.clear();
+}
+
+/* Loads a supervisor into the database, returning true on success and false on
+ * failure. If there is a failure, errorMessage is written with the contents of
+ * the error. */
+bool SuperDB::load_supervisor(std::string appName, std::string path,
+                              std::string* errorMessage)
+{
+    /* Firstly, check whether or not a supervisor has already been loaded for
+     * this application. */
+    SuperIt superIt = supervisors.find(appName);
+    if (superIt != supervisors.end())
+    {
+        *errorMessage = dformat(
+            "A supervisor has already been loaded to application '%s'.",
+            appName.c_str());
+        return false;
+    }
+
+    /* Otherwise, load up. */
+    supervisors[appName] = new SuperHolder(path);
+
+    /* Check for errors as per the specification... */
+    if (supervisors.find(appName)->second->error)
+    {
+        *errorMessage = dlerror();
+        return false;
+    }
+
+    return true;
+}
+
+/* Calls the supervisor for a given application using its entry point. */
+int SuperDB::call_supervisor(std::string appName, PMsg_p* inputMessage,
+                             PMsg_p* outputMessage)
+{
+    SuperIt superFinder = supervisors.find(appName);
+    if (superFinder == supervisors.end()) return -2;
+    /* I know this is hideous, but that's function pointers for you. */
```

Comment:
I'm not sure what you mean by "overwriting a method definition" (given that there are multiple applications). I think that, regardless of what you do, it'll always look horrible *somewhere* because the function pointers are introduced by the SO API.

A "seasoned" C++ developer will recognise what this line is trying to do. An "unseasoned" developer won't, unless they're told that `(*(superFinder->second->entryPoint))` is a function pointer (which is why I added the comment ;)).

Reply:
I was coming from the standpoint of having the definition for a default method that can then be overridden at compile time by a user-defined method.

Function pointers (and virtual functions) are nice and all, but they come with overhead. templates are another way.

In C (under gcc at least, and in c++) you can use a "weak" keyword to indicate weak linkage, so that if the linker comes across a "strong" (in otherwords no keyword) definition of a symbol/method, it uses that instead.

I am not aware of a portable way to do this though

Reply:
The user currently defines the method in the XML, as it's part of application logic. That fragment of the XML is compiled into the shared object, from which this method is loaded.

Why would a user need to define the method at compile time?

Reply:
Ah, sorry, I think I have put this set of comments in the wrong place. I realised this was needed for the supervisor call but I thought there was another place you were allowing overloadable methods at compile time, I am probably being dense